### ISSUE TITLE
risc-v support

### ISSUE THREAD

**Issue by lu-zero** [rr-debugger/rr#3796]
Since there isn't an issue already open, would be possible to list which would be the requirements to have rr working on risc-v?

**Comment by rocallahan** [comment]
The most important requirement is that the CPU needs a hardware performance counter that counts the number of userspace retired instructions, branches, or conditional branches with perfect accuracy (any one of those counter types will do, in that order of preference). I know these counters exist on many RISC-V implementations, but I don't know if they're accurate enough. For example on other CPUs it is common for these counters to be spuriously incremented (e.g. sometimes an interrupt will increment the count of retired branches, or sometimes a speculatively executed instruction that does not retire is mistakenly counted as retired). This would need to be carefully tested and possibly bugs fixed.

Another requirement is the ability to interrupt the CPU when a hardware performance counter reaches a certain value and have the interrupt be delivered within a bounded number of instructions after the counter reached the value.

Naturally we'll need these to be hooked up to the Linux perf-event API. See https://github.com/rr-debugger/rr/blob/master/src/counters-test/README.md for some code that can be adapted to test these things.

Another requirement is that we need to be able to trap on reads from the clock cycle counter, and we'll need Linux kernel API for this like `PR_SET_TSC` does for x86.

**Comment by rocallahan** [comment]
If these basics are present then it can probably be made to work, but the implementation will still be quite a lot of difficult work.

**Comment by lu-zero** [comment]
Thank you!

**Comment by rocallahan** [comment]
One other thing that @khuey reminded me of: RISC-V depends on LL/SC instructions for atomic compare and swap. That's a problem for rr. The ideal solution here would be for the CPU to be configurable to trap on a failed SC, with Linux kernel API for this.

**Comment by rocallahan** [comment]
Yes, if all userspace is compiled to use only the atomic instructions and no LL/SC.

**Comment by Keno** [comment]
If you have more control over the CPU architecture there's other options. For example, when looking at ppc64, we considered the possibility of taking a synchronous trap on SC abort or rewriting the SC instruction to a regular store in microcode.

**Comment by rmacnak-google** [comment]
> ability to interrupt the CPU when a hardware performance counter reaches a certain value

The Sdtrig extension includes an instruction count trigger: https://docs.riscv.org/reference/debug/v1.0/Sdtrig.html#csr-icount

**Comment by rocallahan** [comment]
> This explicitly includes traps taken due to interrupts.
which is unfortunate but manageable. So that's a step forward. Still a lot of work.
